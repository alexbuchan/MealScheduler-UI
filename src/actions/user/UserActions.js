require('es6-promise').polyfill();
const request = require('axios');

import ActionDispatch from './ActionDispatch';
import FlashMessageActions from '../FlashMessageActions/FlashMessageActions';
import ActionsHelper from '../ActionsHelper';
import JWT from '../../lib/JWT/JWT';

class UserActions {
  registerUser = async user => {
    const _endpoint = 'http://localhost:3000/signup';

    let error, response;
    [error, response] = await ActionsHelper.asyncHelper(
      request.post(_endpoint, user)
    );
    
    if (error) {
      FlashMessageActions.dispatchErrorMessage(error.response);
    } else {
      this.handleSignupLoginResponse(response, ActionDispatch.dispatchRegisterUser);
    }
  }

  loginUser = async user => {
    const _endpoint = 'http://localhost:3000/login';

    let error, response;
    [error, response] = await ActionsHelper.asyncHelper(
      request.post(_endpoint, user)
    );
    
    if (error) {
      FlashMessageActions.dispatchErrorMessage(error.response);
    } else {
      this.handleSignupLoginResponse(response, ActionDispatch.dispatchLoginUser);
    }
  }

  logoutUser = () => {
    ActionsHelper.removeCookie();
    ActionDispatch.dispatchLogoutUser();
  }

  retrieveUserDataOnRefresh = () => {
    const cookie = ActionsHelper.getCookie('user');
    if (cookie) {
      const data = JWT.decodeJWTToken(cookie);
      ActionDispatch.dispatchUserDataOnRefresh(data);
    }
  }

  handleSignupLoginResponse = (response, dispatchFunction) => {
    if (response.status === 200) {
      const data = JWT.decodeJWTToken(response.data.token);
      ActionsHelper.setCookie('user', response.data.token);

      dispatchFunction(data);
    }
  }
}

export default new UserActions();
