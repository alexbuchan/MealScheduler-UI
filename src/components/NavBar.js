import React from 'react';

class NavBar extends React.Component {
  render() {
    return (
      <nav className="navbar navbar-expand-lg navbar-dark bg-dark">
        <a className="navbar-brand" href="#">Goldenspear Contacts App</a>
        <button className="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation">
          <span className="navbar-toggler-icon"></span>
        </button>
        <div className="collapse navbar-collapse justify-content-sm-end" id="navbarNav">
          <ul className="navbar-nav">
            <li className="nav-item">
              <a className="nav-link" href="/contacts">Contacts</a>
            </li>
            <li className="nav-item">
              <a className="nav-link" href="/">Signup</a>
            </li>
            <li className="nav-item">
              <a className="nav-link" href="/login">Login <span className="sr-only">(current)</span></a>
            </li>
          </ul>
        </div>
      </nav>
    );
  }
}

export default NavBar;
