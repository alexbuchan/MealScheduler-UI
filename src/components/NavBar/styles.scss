.navbar-height {
  height: $navbarHeight;
  padding-top: $navbarPaddingTop;
  padding-right: $navbarPaddingRight;
}

@media only screen and (max-width: 649px) {
  .navbar-brand {
    font-size: medium;
  }
}