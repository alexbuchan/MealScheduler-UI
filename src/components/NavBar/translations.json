{
  "en": {
    "navbar": {
      "schedule": "Schedule",
      "recipes": "Recipes",
      "settings": "Settings",
      "logout": "Log out"
    }
  },

  "es": {
    "navbar": {
      "schedule": "Calendario",
      "recipes": "Recetas",
      "settings": "Ajustes",
      "logout": "Cerrar Sesión"
    }
  },

  "fr": {
    "navbar": {
      "schedule": "Programme",
      "recipes": "Recette",
      "settings": "Paramètres",
      "logout": "Se déconnecter"
    }
  },

  "cat": {
    "navbar": {
      "schedule": "Horari",
      "recipes": "Recetes",
      "settings": "Configuració",
      "logout": "Tancar Sessió"
    }
  }
}