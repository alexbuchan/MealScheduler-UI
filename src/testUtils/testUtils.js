import React from 'react';
import { shallow } from 'enzyme';

export const componentSetup = (Component, props = {}) => {
  const instance = shallow(<Component { ...props } />);
  return instance;
}

export const findByTestAttribute = (instance, attribute) => {
  const wrapper = instance.find(`[data-test="${attribute}"]`);
  return wrapper;
}