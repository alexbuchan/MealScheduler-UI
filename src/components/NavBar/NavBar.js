import React from 'react';
import { Link } from "react-router-dom";
import UserActions from '../../actions/user/UserActions';
import UserStore from '../../stores/UserStore/UserStore';

class NavBar extends React.Component  {
  constructor() {
    super();

    this.state = {
      user: UserStore.getUserState().user
    }

    this._onChange = this._onChange.bind(this);
  }

  handleLogout = (ev) => {
    UserActions.logoutUser();
  }

  _onChange() {
    this.setState({
      user: UserStore.getUserState().user,
    });
  }

  componentDidMount() {
    UserStore.addChangeListener(this._onChange);
  }

  componentWillUnmount() {
    UserStore.removeChangeListener(this._onChange);
  }

  renderUsername = () => {
    if (this.state.user) {
      return (
        <li className="nav-item mobile-nav-item">
          <span className="nav-link user-profile" to="/schedule">{ this.state.user.username }<span className="sr-only">(current)</span></span>
        </li>
      );
    }

    return null;
  }

  renderSignup = () => {
    if (!this.state.user) {
      return (
        <li className="nav-item mobile-nav-item">
          <Link className="nav-link" to="/">Signup</Link>
        </li>
      );
    }

    return null;
  }

  renderLogin = () => {
    if (this.state.user) {
      return (
        <li className="nav-item mobile-nav-item">
          <Link
            onClick={ this.handleLogout }
            className="nav-link"
            data-test='logout-nav-link'
            to="/login"
          >
            Log out
            <span className="sr-only">(current)</span>
          </Link>
        </li>
      );
    }

    return (
      <li className="nav-item mobile-nav-item">
        <Link className="nav-link" to="/login">Login<span className="sr-only">(current)</span></Link>
      </li>
    );
  }

  renderSettings = () => {
    if (this.state.user) {
      return (
        <li className="nav-item mobile-nav-item">
          <Link className="nav-link" data-test='settings-nav-link' to="/settings">Settings<span className="sr-only">(current)</span></Link>
        </li>
      );
    }

    return null;
  }

  render() {
    return (
      <nav className="navbar-height navbar navbar-expand-xs navbar-expand-sm navbar-expand-md navbar-expand-lg navbar-dark bg-dark">
        <Link className="navbar-brand" to="">Meal Scheduler</Link>
        <button className="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation">
          <span className="navbar-toggler-icon"></span>
        </button>
        <div className="collapse navbar-collapse justify-content-sm-end" id="navbarNav">
          <ul className="navbar-nav">
            { this.renderUsername() }
            <li className="nav-item mobile-nav-item">
              <Link className="nav-link" to="/schedule">Schedule</Link>
            </li>

            { this.renderSettings() }
            { this.renderSignup() }
            { this.renderLogin() }
          </ul>
        </div>
      </nav>
    );
  }
}

export default NavBar;
