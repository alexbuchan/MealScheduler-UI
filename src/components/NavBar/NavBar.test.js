import NavBar from './NavBar';
import UserActions from '../../actions/user/UserActions';
import { componentSetup, findByTestAttribute } from '../../testUtils/testUtils';

jest.mock('../../actions/user/UserActions');

describe('NavBar', () => {
  describe('When user is not logged in', () => {
    let instance;

    beforeEach(() => {
      instance = componentSetup(NavBar);
    });

    it('Should render contacts, login, and signup in the navbar', () => {
      expect(instance).toMatchSnapshot();
    });
  });

  describe('When user is logged in', () => {
    let instance;
    let props = {
      user: {
        username: 'username'
      }
    }

    beforeEach(() => {
      instance = componentSetup(NavBar, props);
    });

    it('Should render username, contacts, and logout in the navbar', () => {
      expect(instance).toMatchSnapshot();
    });

    describe('#handleLogout', () => {
      it('calls the user logout action', () => {
        const logoutLink = findByTestAttribute(instance, 'logout-nav-link');
        logoutLink.simulate('click', { button: 0 });

        expect(UserActions.logoutUser).toHaveBeenCalledTimes(1);
      });
    });
  });
});