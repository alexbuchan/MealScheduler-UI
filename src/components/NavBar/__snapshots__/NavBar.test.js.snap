// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar When user is logged in Should render username, contacts, and logout in the navbar 1`] = `
<nav
  className="navbar-height navbar navbar-expand-xs navbar-expand-sm navbar-expand-md navbar-expand-lg navbar-dark bg-dark"
>
  <Link
    className="navbar-brand"
    to=""
  >
    Signup and Login App
  </Link>
  <button
    aria-controls="navbarNav"
    aria-expanded="false"
    aria-label="Toggle navigation"
    className="navbar-toggler"
    data-target="#navbarNav"
    data-toggle="collapse"
    type="button"
  >
    <span
      className="navbar-toggler-icon"
    />
  </button>
  <div
    className="collapse navbar-collapse justify-content-sm-end"
    id="navbarNav"
  >
    <ul
      className="navbar-nav"
    >
      <li
        className="nav-item"
      >
        <Link
          className="nav-link"
          to="/contacts"
        >
          username
          <span
            className="sr-only"
          >
            (current)
          </span>
        </Link>
      </li>
      <li
        className="nav-item"
      >
        <Link
          className="nav-link"
          to="/contacts"
        >
          Contacts
        </Link>
      </li>
      <li
        className="nav-item"
      >
        <Link
          className="nav-link"
          data-test="logout-nav-link"
          onClick={[Function]}
          to="/login"
        >
          Log out
          <span
            className="sr-only"
          >
            (current)
          </span>
        </Link>
      </li>
    </ul>
  </div>
</nav>
`;

exports[`NavBar When user is not logged in Should render contacts, login, and signup in the navbar 1`] = `
<nav
  className="navbar-height navbar navbar-expand-xs navbar-expand-sm navbar-expand-md navbar-expand-lg navbar-dark bg-dark"
>
  <Link
    className="navbar-brand"
    to=""
  >
    Signup and Login App
  </Link>
  <button
    aria-controls="navbarNav"
    aria-expanded="false"
    aria-label="Toggle navigation"
    className="navbar-toggler"
    data-target="#navbarNav"
    data-toggle="collapse"
    type="button"
  >
    <span
      className="navbar-toggler-icon"
    />
  </button>
  <div
    className="collapse navbar-collapse justify-content-sm-end"
    id="navbarNav"
  >
    <ul
      className="navbar-nav"
    >
      <li
        className="nav-item"
      >
        <Link
          className="nav-link"
          to="/contacts"
        >
          Contacts
        </Link>
      </li>
      <li
        className="nav-item"
      >
        <Link
          className="nav-link"
          to="/"
        >
          Signup
        </Link>
      </li>
      <li
        className="nav-item"
      >
        <Link
          className="nav-link"
          to="/login"
        >
          Login
          <span
            className="sr-only"
          >
            (current)
          </span>
        </Link>
      </li>
    </ul>
  </div>
</nav>
`;
